docs/reference/src/main/asciidoc/contexts.asciidoc

```diff
@@ -250,3 +250,165 @@ conversation context, as can the current conversation.
+        Callable<T> wrappedTask = new Callable<T>() {
+
+            @Override
+            public T call() throws Exception {
+                // Get WeldManager,get instances of @Bound contexts for request, session and conversation scopes
+                WeldManager weldManager = CDI.current().select(WeldManager.class).get();
+                BoundRequestContext requestContext = weldManager.instance().select(BoundRequestContext.class, BoundLiteral.INSTANCE).get();
+                BoundSessionContext sessionContext = weldManager.instance().select(BoundSessionContext.class, BoundLiteral.INSTANCE).get();
+                BoundConversationContext conversationContext = weldManager.instance().select(BoundConversationContext.class, BoundLiteral.INSTANCE).get();
+
+                // We will be using bound contexts, prepare backing structures for contexts
+                Map<String, Object> sessionMap = new HashMap<>();
+                Map<String, Object> requestMap = new HashMap<>();
+                BoundRequest boundRequest = new MutableBoundRequest(requestMap, sessionMap);
+
+                // activate contexts
+                requestContext.associate(requestMap);
+                requestContext.activate();
+                sessionContext.associate(sessionMap);
+                sessionContext.activate();
+                conversationContext.associate(boundRequest);
+                conversationContext.activate();
+
+                // propagate all contexts that have some bean in them
+                if (scopeToContextualInstances.get(requestContext.getScope()) != null) {
+                    requestContext.clearAndSet(scopeToContextualInstances.get(requestContext.getScope()));
+                }
+                if (scopeToContextualInstances.get(sessionContext.getScope()) != null) {
+                    sessionContext.clearAndSet(scopeToContextualInstances.get(sessionContext.getScope()));
+                }
+                if (scopeToContextualInstances.get(conversationContext.getScope()) != null) {
+                    conversationContext.clearAndSet(scopeToContextualInstances.get(conversationContext.getScope()));
+                }
+
+                // now execute the actual original task
+                T result = task.call();
+
+                // cleanup, context deactivation, do not trigger @PreDestroy/@Disposes
+                requestContext.deactivate();
+                conversationContext.deactivate();
+                sessionContext.deactivate();
+
+                // all done, return
+                return result;
+            }
+        };
+        return executor.submit(wrappedTask);
+    }
+}
+----------------------------------------------------------------------------------------------------------------------------------------------------------
+
+==== Pitfalls and drawbacks
+
+There are several things that can possibly go wrong when propagating contexts.
```

Comment:
This is a long list ;-)

Reply:
As a matter of fact, I was just warming up... ;)
Got better ideas? I wanted to show that this is not the ultimate tool and you pay a price for using it.